src/layers/transformer.rs

```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
+        TransformerDropouts {
+            attn_output_dropout: Box::new(Identity),
+            ffn_output_dropout: Box::new(Identity),
+            parallel_attn_dropout: Box::new(Dropout::new(p)),
+        }
+    }
+}
+
+/// Transformer layer.
+///
+/// This is a generic transformer layer that is used by `DecoderLayer` and
+/// `EncoderLayer` to provide specialized layers.
+///
+/// See [Vaswani et al. (2017)](https://arxiv.org/abs/1706.03762).
+struct TransformerLayer {
+    use_parallel_attention: bool,
+    mha: SelfAttention,
+    ffn: PointwiseFeedForward,
+    dropouts: TransformerDropouts,
+    norms: TransformersLayerNorms,
+}
+
+impl TransformerLayer {
+    /// Construct a transformer layer.
+    ///
+    /// * `attention_layer` - The attention layer to use in the transformer
+    ///   layers.
+    /// * `feed_forward_layer` - The feed-forward layer to use in the
+    ///   transformer layers.
+    /// * `dropouts` - The dropouts to use in the transformer layers.
+    /// * `layer_norms` - The layer norms to use in the transformer
+    /// * `use_parallel_attention` - Use parallel attention.
+    fn new(
+        attention_layer: SelfAttention,
+        feed_forward_layer: PointwiseFeedForward,
+        dropouts: TransformerDropouts,
+        layer_norms: TransformersLayerNorms,
+        use_parallel_attention: bool,
+    ) -> Self {
+        Self {
+            dropouts,
+            ffn: feed_forward_layer,
+            mha: attention_layer,
+            norms: layer_norms,
+            use_parallel_attention,
+        }
+    }
+
+    /// Apply the transformer layer to the given piece hidden representations.
+    ///
+    /// * `input` - Hidden representations to apply the layer to.
+    ///   *Shape:* `(batch_size, seq_len, width)`
+    /// * `attention_mask` - Attention mask. Sequence elements for which the
+    ///    corresponding mask element is set to `False` are ignored
+    ///    during attention calculation.
+    /// * `use_causal_mask` - Mask out succeeding sequence elements when `True`.
+    /// * `cache` - Key/value cache to avoid recomputing key/value representations
+    ///    for tokens that were previously seen.
+    /// * `positions` - Input positions. Positions are needed to look up rotary
+    ///    embeddings. Normally, these positions are calculated automatically.
+    ///    But if the positions deviate for some reason, they can be provided
+    ///    through this argument.
+    ///    *Shape:* `(batch_size, seq_len)`
+    /// * `store_cache` - Whether to cache the key/value representations for
+    ///   future reuse.
+    ///
+    /// Returns layer output and the key/value cache.
+    /// *Shape:* ``(batch_size, seq_len, width)``
+    #[allow(clippy::too_many_arguments)]
+    fn forward(
+        &self,
+        input: &Tensor,
+        train: bool,
+        attention_mask: &AttentionMask,
+        _cache: Option<&KeyValueCache>,
+        _positions: Option<&Tensor>,
+        _store_cache: bool,
```

Comment:
```suggestion
        attention_mask: &AttentionMask,
        _cache: Option<&KeyValueCache>,
        _positions: Option<&Tensor>,
        _store_cache: bool,
        train: bool, 
```
Unless we want to make `train` special (since it's also exposed by `ModuleT`) and keep it as the second argument?

Reply:
Yeah, that was my motivation to move it to the second position. Not sure though, we need to come up with a sensible ordering, maybe something like:

* inputs/xs
* Other input arguments in alphabetical order (attention mask, positions, cache).
* Boolean/option arguments in alphabetical order.

The only thing I don't really like is that an argument which is pretty much mandatory (train) ends up shifting its position all the time. On the other hand, reserving a position for it will only make the ordering rules more complex.

Reply:
The above order sounds good to me!